solidity/SolidityParser.g4

```diff
@@ -0,0 +1,861 @@
+ */
+unicodeStringLiteral
+    : UnicodeStringLiteral+
+    ;
+
+/**
+ * Number literals can be decimal or hexadecimal numbers with an optional unit.
+ */
+numberLiteral
+    : DecimalNumber
+    | HexNumber
+    ;
+
+/**
+ * A curly-braced block of statements. Opens its own scope.
+ */
+block
+    : LBrace (statement | uncheckedBlock)* RBrace
+    ;
+
+uncheckedBlock
+    : Unchecked block
+    ;
+
+statement
+    : block
+    | simpleStatement
+    | ifStatement
+    | forStatement
+    | whileStatement
+    | doWhileStatement
+    | continueStatement
+    | breakStatement
+    | tryStatement
+    | returnStatement
+    | emitStatement
+    | revertStatement
+    | assemblyStatement
+    ;
+
+//@doc:inline
+simpleStatement
+    : variableDeclarationStatement
+    | expressionStatement
+    ;
+
+/**
+ * If statement with optional else part.
+ */
+ifStatement
+    : If LParen expression RParen statement (Else statement)?
+    ;
+
+/**
+ * For statement with optional init, condition and post-loop part.
+ */
+forStatement
+    : For LParen (simpleStatement | Semicolon) (expressionStatement | Semicolon) expression? RParen statement
+    ;
+
+whileStatement
+    : While LParen expression RParen statement
+    ;
+
+doWhileStatement
+    : Do statement While LParen expression RParen Semicolon
+    ;
+
+/**
+ * A continue statement. Only allowed inside for, while or do-while loops.
+ */
+continueStatement
+    : Continue Semicolon
+    ;
+
+/**
+ * A break statement. Only allowed inside for, while or do-while loops.
+ */
+breakStatement
+    : Break Semicolon
+    ;
+
+/**
+ * A try statement. The contained expression needs to be an external function call or a contract creation.
+ */
+tryStatement
+    : Try expression (Returns LParen returnParameters = parameterList RParen)? block catchClause+
+    ;
+
+/**
+ * The catch clause of a try statement.
+ */
+catchClause
+    : Catch (identifier? LParen (arguments = parameterList) RParen)? block
+    ;
+
+returnStatement
+    : Return expression? Semicolon
+    ;
+
+/**
+ * An emit statement. The contained expression needs to refer to an event.
+ */
+emitStatement
+    : Emit expression callArgumentList Semicolon
+    ;
+
+/**
+ * A revert statement. The contained expression needs to refer to an error.
+ */
+revertStatement
+    : Revert expression callArgumentList Semicolon
+    ;
+
+/**
+ * An inline assembly block.
+ * The contents of an inline assembly block use a separate scanner/lexer, i.e. the set of keywords and
+ * allowed identifiers is different inside an inline assembly block.
+ */
+assemblyStatement
+    : Assembly AssemblyDialect? assemblyFlags? AssemblyLBrace yulStatement* YulRBrace
+    ;
+
+/**
+ * Assembly flags.
+ * Comma-separated list of double-quoted strings as flags.
+ */
+assemblyFlags
+    : AssemblyBlockLParen AssemblyFlagString (AssemblyBlockComma AssemblyFlagString)* AssemblyBlockRParen
+    ;
+
+//@doc:inline
+variableDeclarationList
```

Comment:
Let's keep it, since it is in the original spec.

Reply:
I will create an Issue in the originating repo for this and other issues. There are many useless parentheses, which don't add anything to the clarity of the grammar.

Also, there is ambiguity in `expression,` etc. Ambiguity is concerning because it means this grammar was not thoroughly debugged.